Add Support for Migration Statistics API Call

phip1611 opened the pull request:
## TL;DR

This extends the internal API with a `vm_progress` function and adds a `vm.migration-progress` HTTP endpoint including support in `ch-remote` via `ch-remote migration-process` to query the latest migration progress.

# Motivation

Having live and frequently refreshing statistics/metrics about an
ongoing live migration is especially interesting for debugging and
monitoring, such as checking the actual network throughput. With the
proposed changes, for the first time, we will be able to see how live
migrations behave and create benchmarking infrastructure around it.

The ch driver in libvirt will use these information to populate its
`virsh domjobinfo` information.

# Design

We will add a new API endpoint to query information for ongoing live
migrations. The new endpoint will also serve to query information about
any previously failed or canceled migrations. The SendMigration call
will no longer be blocking (wait until the migration is done) but
instead just dispatch the migration.

This streamlines the behavior with QEMU and simplifies management
software.

When one queries the endpoint, a frequently refreshed snapshot of the
migration statistics and progress will be returned. The data will not be
assembled on the fly.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>


## Steps to Merge
- [x] test it locally using `ch-remote`
- [x] add libvirt-tests testcase and verify everything works
- [x] finish refactoring for "non-blocking send-migration" (2026-02-03)
- [x] deploy it on a node in SAP land and see if it works 
- [x] merge this when we know the libvirt part is also fine (I however think that this is 99% already)
- [x] wait for the latest libvirt/libvirt-tests fixes

phip1611 commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,257 @@
+        /// The final memory transmission info.
+        memory_info: MemoryTransmissionInfo,
+    },
+    Failed {
```
currently I don't like that this phase is part of this enum. Instead, on the top level we should have

`Ongoing(InnerStateA), Cancelled(InnerStateB), Failed(InnerStateC)`

tpressure commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,257 @@
+    pub memory_bytes_transmitted: u64,
+    pub memory_pages_4k_transmitted: u64,
+    pub memory_bytes_remaining_iteration: u64,
+    pub memory_bytes_remaining_total: u64,
```
qemu also reports the dirty rate. Do we want to report this too?

tpressure replied:
what's also missing is some sort of status (migration in-progess, migration-finished, ...)

phip1611 replied:
> emu also reports the dirty rate. Do we want to report this too?

100%, sure. I just forgot it here

> migration in-progess, migration-finished

I have it already, but I want to refactor it to make it more promiment. See my comment https://github.com/cyberus-technology/cloud-hypervisor/pull/43#pullrequestreview-3517985531

phip1611 commented on src/bin/ch-remote.rs:
```diff
@@ -301,6 +302,21 @@ fn rest_api_do_command(matches: &ArgMatches, socket: &mut UnixStream) -> ApiResu
+            let clear = matches
+                .subcommand_matches("migration-progress")
+                .unwrap()
+                .get_one::<bool>("clear")
```
I think we can completely nuke that `clear` complexity - there is no problem with keeping the old state around. As soon as a new migration starts, the state is overwritten anyway

phip1611 commented on vmm/src/api/mod.rs:
```diff
@@ -296,6 +301,13 @@ pub struct VmSendMigrationData {
 }
 
+#[derive(Clone, Deserialize, Serialize, Debug)]
+pub struct VmMigrationProgressData {
```
I think we can completely nuke that `clear` complexity - there is no problem with keeping the old state around. As soon as a new migration starts, the state is overwritten anyway

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -2336,10 +2403,26 @@ impl Vmm {
+        vm.complete_migration()?;
+
+        // Give management software a chance to fetch the migration state.
+        info!("Sleeping five seconds before shutting off.");
```
TODO remove this before merge

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -2336,10 +2403,26 @@ impl Vmm {
+        info!("Sleeping five seconds before shutting off.");
+        // TODO right now, the http-server is single-threaded and the blocking
+        // start-migration API call will block other requests here.
+        thread::sleep(Duration::from_secs(5));
```
TODO remove this before merge

phip1611 committed:
vmm: add migration-progress HTTP endpoint

This is part of the commit series to enable live updates about an
ongoing live migration. See the first commit for an introduction.

In this commit, we add the HTTP endpoint to export ongoing VM
live-migration progress.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,346 @@
+/// [live-migration protocol]: super::protocol
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct MigrationProgressAndStatus {
+    /// UNIX timestamp of the start of the live-migration process.
```
please note that this structure will be public API for ever and once we reploy it, it will be hard to change

phip1611 commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,346 @@
+///
+/// [live-migration protocol]: super::protocol
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct MigrationProgressAndStatus {
```
should I add a `version` parameter here? Similar to live migration which also has (or will have?) a version?

olivereanderson replied:
I think that would be reasonable.

phip1611 replied:
we discussed this and so far we think this is not needed

phip1611 committed:
vm-migration: prepare progress types for new API endpoint

This is the first commit in a series of commits to introduce a new API
endpoint in Cloud Hypervisor to report progress and live-insights about
an ongoing live migration.

# Motivation

Having live and frequently refreshing statistics/metrics about an
ongoing live migration is especially interesting for debugging and
monitoring, such as checking the actual network throughput. With the
proposed changes, for the first time, we will be able to see how live
migrations behave and create benchmarking infrastructure around it.

The ch driver in libvirt will use these information to populate its
`virsh domjobinfo` information.

# Design

We will add a new API endpoint to query information for ongoing live
migrations. The new endpoint will also serve to query information about
any previously failed or canceled migrations. The SendMigration call
will no longer be blocking (wait until the migration is done) but
instead just dispatch the migration.

This streamlines the behavior with QEMU and simplifies management
software.

When one queries the endpoint, a frequently refreshed snapshot of the
migration statistics and progress will be returned. The data will not be
assembled on the fly.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
vmm: add migration-progress API endpoint

This is part of the commit series to enable live updates about an
ongoing live migration. See the first commit for an introduction.

We decided to use an Option<> rather than a Result<> as
there isn't really an error that can happen when we query this endpoint.
A previous snapshot may either be there or not. It also doesn't make
sense here to check if the current VM is running, as users should always
be able to query information about the past (failed or canceled) live
migration.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+    /// The total size of transmitted bytes.
+    pub memory_bytes_transmitted: u64,
+    /// The amount of remaining bytes for this iteration.
+    pub memory_bytes_remaining_iteration: u64,
+    /// The amount of transmitted 4k pages.
+    pub memory_pages_4k_transmitted: u64,
+    /// The amount of remaining 4k pages for this iteration.
+    pub memory_pages_4k_remaining_iteration: u64,
```
Question: What about larger pages?

phip1611 replied:
VMM + KVM transparently splits huge-pages for a migration. I'm however not sure if this also applies to read-only pages. We should discuss this again with Thomas

phip1611 replied:
From offline discussion. KVM splits all pages to 4k for the scope of the migration

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+        /// Percent in range `0..=100`.
+        vcpu_throttle_percent: u8,
+    },
+}
```
I see that every variant has a field of type `MemoryTransmissionInfo`. Maybe it would make sense to take that out of the enum?
You could do that either with a wrapper struct, or just work with a tuple e.g.: 
```
pub type MigrationProgressInfo = (MigrationProgressState, MigrationProgressInfo)`;
```

phip1611 replied:
Let me think about this again!

phip1611 replied:
refactored

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+                ..
+            } => *memory_transmission_info,
+        }
+    }
```
See my comment above with regards to taking `MemoryTransmissionInfo` out of the enum.

phip1611 replied:
Let me think about this again! I iterates multiple times over the design. It might also makes sense to move that out of the enum at this place, yes

phip1611 replied:
refactored

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+                ..
+            } => Some(*vcpu_throttle_percent),
+            _ => None,
+        }
```
I think `if let` would be more idiomatic here. Something like this:
```suggestion
if let Self::Ongoing{vcpu_throttle_percent, ..} = self {
     Some(*vcpu_throttle_percent)
} else {
      None
}
```

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+fn current_unix_timestamp_ms() -> u64 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
```
Consider using `expect` instead of `unwrap` in this case.

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+        }
+    }
+
+    /// Updates the state of an ongoing migration.
```
Consider adding a  `# Panics` section here.

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+    /// Sets the underlying state to [`MigrationProgressState::Cancelled`] and
+    /// updates all corresponding metadata.
+    ///
+    /// After this state change, the object is supposed to be handled as immutable.
```
Consider adding a `# Panics` section here too.

phip1611 replied:
thanks, fixed!

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+            );
+        }
+        self.timestamp_snapshot = current_unix_timestamp_ms();
+        self.timestamp_snapshot = current_unix_timestamp_ms();
```
Why is this doubled?

phip1611 replied:
fixed, thanks

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+    /// Sets the underlying state to [`MigrationProgressState::Failed`] and
+    /// updates all corresponding metadata.
+    ///
+    /// After this state change, the object is supposed to be handled as immutable.
```
Again consider adding a `# Panics` section.

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+    /// Sets the underlying state to [`MigrationProgressState::Finished`] and
+    /// updates all corresponding metadata.
+    ///
+    /// After this state change, the object is supposed to be handled as immutable.
```
Again consider adding a `# Panics` section.

phip1611 replied:
fixed

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+
+    // Helpful to see what the API will look like.
+    #[test]
+    fn print_json() {
```
Is this test a leftover from development? I don't see any asserts here.

olivereanderson replied:
@phip1611 ?

phip1611 replied:
Fair question! I think I can remove this in favor a real unit test

olivereanderson commented on vmm/src/api/mod.rs:
The commit message contains:
> an `vm_migration` API call

but it should be "a `vm_migration` API call".

phip1611 replied:
next time, please write the title of the commit message to the comment. I have no chance to find out which commit message this comment belongs to other than brute force :)

olivereanderson commented on vmm/src/api/mod.rs:
```diff
@@ -314,6 +319,9 @@ pub enum ApiResponsePayload {
     VmInfo(VmInfoResponse),
 
+    /// The progress of a possibly ongoing live migration.
+    VmMigrationProgress(Box<Option<MigrationProgressAndStatus>>),
```
I guess the `None` variant will only ever be present if the API is called before a live migration has started?

phip1611 replied:
yes

phip1611 replied:
I think about refactoring the whole thing to `Result<MigrationProgress>` and drop the Option

phip1611 replied:
Hm I think Option makes the most sense as is the most consistent API without suddenly changing semantics

olivereanderson commented on vmm/src/api/http/http_endpoint.rs:
```diff
@@ -708,6 +709,32 @@ impl EndpointHandler for VmmShutdown {
+            {
+                Ok(info) => {
+                    let mut response = Response::new(Version::Http11, StatusCode::OK);
+                    let info_serialized = serde_json::to_string(&info).unwrap();
```
Nit: It would be good to either replace `unwrap` with `expect` or to add a comment explaining why it is OK to unwrap in this case.

phip1611 replied:
I'm not happy with it but it is done like this everywhere in ch-remote. I'd much rather change this in upstream CHV

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2207,6 +2243,34 @@ impl Vmm {
+                    MigrationProgressState::Failed { .. } => {}
+                    MigrationProgressState::Finished { .. } => {}
+                }
+            }
```
When would you expect this to match the `Some` variant? Isn't this right at the begging of the migration? I would expect it to be unconditionally `None` here.

phip1611 replied:
> it to be unconditionally None here.

Nope. when a migration failed or was cancelled (which is not yet supported but might be eventually), the snapshot will be populated with the stats of the previous run

olivereanderson replied:
Ah, of course. On the sender side you are allowed to retry migrations, just not on the receiver side. Makes sense. Thanks!

olivereanderson commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,366 @@
+    /// as all bytes have on fixed value (e.g., a zero page).
+    pub memory_pages_constant_count: u64,
+    /// Current memory dirty rate in pages per seconds.
+    pub memory_dirty_rate_pps: u64,
```
Question: Is this also relevant when auto-converged is eventually replaced with post-copy?

phip1611 replied:
Yes as there is a pre-copy phase before one switches to postcopy

olivereanderson reviewed (commented):
Thanks for your work on this!

I wish the PR description, or some other form of documentation, explained the intended use cases for this feature.

Is the main use case debugging slow/hanging live migrations, or is this intended to be used to collect metrics for every live migration which will then be used to provide statistical insights? If it is the latter, it would be good to document (somewhere) the compatibility story with standard metrics services like Prometheus. I think @snue kowns a thing or two about such topics :slightly_smiling_face: 

The [metrics crate](https://crates.io/crates/metrics) seems to be the metrics analogue of `log`/`tracing` and it would be good to know why that isn't a good fit for the task(s) addressed by this PR. I am assuming that there is some context I am missing and that it is quite possible that I have some misconceptions about the problems you are trying to address.

Otherwise I would just like to say that the code looks good and it was easy to follow the implementation for the most part.

olivereanderson commented on vmm/src/lib.rs:
```diff
@@ -2205,6 +2243,34 @@ impl Vmm {
+                    MigrationProgressState::Failed { .. } => {}
+                    MigrationProgressState::Finished { .. } => {}
+                }
+            }
```
Nit: Something like this might be simpler:
```suggestion
if lock.as_ref().is_some_and(|snapshot| matches!(snapshot, MigrationProgressState::Ongoing{..}) {
    // if this fails, we made a programming error in our state handling
    panic!("migration already ongoing");
}
```

phip1611 replied:
hm, and how about

```rust
            if let Some(snapshot) = lock.as_ref() {
                match snapshot.state {
                    MigrationProgressState::Ongoing { .. } => {
                        // if this fails, we made a programming error in our state handling
                        panic!("migration already ongoing");
                    }
                    _ => {}
                }
            }
```

as it is the more readable in my opinion.

or 

```rust
            if let Some(snapshot) = lock.as_ref() && let MigrationProgressState::Ongoing { .. }  = snapshot.state {
                // if this fails, we made a programming error in our state handling
                        panic!("migration already ongoing");
            }
```

phip1611 replied:
I decided for

```
            match lock.as_ref() {
                Some(MigrationState::Ongoing { .. }) => {
                    // if this fails, we made a programming error in our state handling
                    panic!("migration already ongoing");
                }
                _ => {},
            }
```

olivereanderson replied:
I find
```
if lock.as_ref().is_some_and(|snapshot| matches!(snapshot, MigrationProgressState::Ongoing{..}) {
    // if this fails, we made a programming error in our state handling
    panic!("migration already ongoing");
}
```
The most readable personally, but the let chain is also OK. I find the full `match` version the most confusing, but I don't really have any strong opinion here. Do whatever you think is best :slightly_smiling_face:

olivereanderson replied:
I also now noticed that:
```
    // if this fails, we made a programming error in our state handling
```
Should rather be
```
   // If this matches, then we made a programming error in our state handling
```

phip1611 committed:
vmm: prepare migration code for migration progress reporting

This is part of the commit series to enable live updates about an
ongoing live migration. See the first commit for an introduction.

This commit prepares the avoidance of naming clashes in the following.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
vmm: actually populate migration progress

This is part of the commit series to enable live updates about an
ongoing live migration. See the first commit for an introduction.

This commit actually brings all the functionality together. The first
version has the limitation that we populate the latest snapshot once per
memory iteration, although this is the most interesting part by far. In
a follow-up, we can make this more fine-grained.

We guarantee that as soon as SendMigration returns, migration progress
can be fetched as the underlying data source is populated.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 commented:
> Thanks for your work on this!
> 
> I wish the PR description, or some other form of documentation, explained the intended use cases for this feature.


I've updated the code.

> The [metrics crate](https://crates.io/crates/metrics) seems to be the metrics analogue of log/tracing and it would be good to know why that isn't a good fit for the task(s) addressed by this PR. I am assuming that there is some context I am missing and that it is quite possible that I have some misconceptions about the problems you are trying to address.

I don't see how this would help. These statistics are not "nice helpers that we need sometimes to debug". They are a fundamental and important functionality we want to export via API.

I hope the new PR description makes the intent here clearer.

amphi commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,394 @@
+    pub memory_pages_4k_remaining_iteration: u64,
+    /// The amount of zero pages for that we could take a shortcut
+    /// as all bytes have on fixed value (e.g., a zero page).
+    pub memory_pages_constant_count: u64,
```
This comment seems a bit weird. I guess it should be something like

> The amount of pages that we could take a shortcut for,
> because all bytes have a fixed value (e.g., zero pages).

amphi commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,394 @@
+    /// This follows after a precopy phase.
+    ///
+    /// Not used for local migrations.
+    MemoryPostcopy,*/
```
Is it intended that these are comments in a comment?

amphi commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,394 @@
+    }
+}
+
+/// The different state of a migration, covering steady progress and failure.
```
`The different stateS ...`?

phip1611 replied:
fixed

amphi commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,394 @@
+            transportation_mode,
+            state: MigrationState::Ongoing {
+                phase: OngoingMigrationPhase::Starting,
+                vcpu_throttle_percent: 42,
```
Is the 42 intended?

phip1611 replied:
fixed, thanks!

amphi commented on vmm/src/lib.rs:
```diff
@@ -2346,10 +2424,26 @@ impl Vmm {
+        // Give management software a chance to fetch the migration state.
+        // The VMM already executes on the other side and keeping Cloud Hypervisor running for a
+        // couple of more seconds is fine.
+        info!("Sleeping five seconds before shutting off.");
```
I think this waiting, and maybe also the `mark_as_finished()` a few lines above, should happen in `check_migration_result`. Otherwise the management software would see the `mark_as_failed` very late.

phip1611 replied:
Agree about the sleep. Not sure about the other. Other thoughts?

phip1611 replied:
Update. all of this need to happen in the end of the migration thread, not in `check_migration_result`. Otherwise, the API thread is blocked and API users can never query the latest state.

phip1611 replied:
Update: we decided to remove the sleep. Instead, there is now a `keep_alive` flag for SendMigration

amphi reviewed (commented):
Looks good, just a few small comments and nits.

phip1611 committed:
vmm: improve logging of ongoing live migrations

The logging is not very spammy nor costly (iterations take seconds to
dozens of minutes) and is clearly a win for us to debug things.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

amphi commented on cloud-hypervisor/src/bin/ch-remote.rs:
```diff
@@ -524,8 +527,65 @@ fn rest_api_do_command(matches: &ArgMatches, socket: &mut UnixStream) -> ApiResu
+
+            // Wait for migration to finish
+            loop {
+                let response = simple_api_full_command_and_response(
```
Do we really want that? As far as I know `ch-remote` is just a tool that makes using the REST API easier, thus I would expect that it behaves like the REST API. If I understand correctly, you don't even give the user the possibility to not see the statistics etc. when using `ch-remote`?

phip1611 replied:
This would have to be discussed with upstream. I have a tendency to handle it this way, as it keep the blocking semantics in ch-remote.

Fair question, tho!

amphi commented on vmm/src/lib.rs:
```diff
@@ -2240,13 +2240,16 @@ impl Vmm {
             s.iteration_duration = s.iteration_start_time.elapsed();
             info!(
-                "iteration:{},cost={}ms,throttle={}%,transmitted={}MiB,dirty_rate={}pps,Mebibyte/s={:.2}",
+                "iter={},dur={}ms,overhead={}ms,throttle={}%,size={}MiB,dirtyrate={},bandwidth={:.2}MiBs,downtime={}ms",
```
nit: missing unit at dirtyrate (I guess its pages per second?)

phip1611 replied:
good catch!

amphi commented on vm-migration/src/lib.rs:
```diff
@@ -9,10 +9,12 @@ use thiserror::Error;
 use crate::protocol::MemoryRangeTable;
 
-mod bitpos_iterator;
+pub mod progress;
```
In commit "vm-migration: prepare progress types for new API endpoint" you say that this needs the API "asynchronization". This isn't true anymore, right?

phip1611 replied:
You are right, I should rephrase the commit history in the early commits as we are now working towards a different goal!

phip1611 replied:
fixed

amphi commented on vmm/src/api/http/http_endpoint.rs:
```diff
@@ -58,9 +58,10 @@ use crate::api::http::http_endpoint::fds_helper::{attach_fds_to_cfg, attach_fds_
-    VmAddUserDevice, VmAddVdpa, VmAddVsock, VmBoot, VmConfig, VmCounters, VmDelete, VmNmi, VmPause,
-    VmPowerButton, VmReboot, VmReceiveMigration, VmReceiveMigrationData, VmRemoveDevice, VmResize,
-    VmResizeDisk, VmResizeZone, VmRestore, VmResume, VmSendMigration, VmShutdown, VmSnapshot,
+    VmAddUserDevice, VmAddVdpa, VmAddVsock, VmBoot, VmConfig, VmCounters, VmDelete,
```
Commit "vmm: add migration-progress HTTP endpoint" again contains outdated information about the async HTTP API.

phip1611 replied:
This is reverted in a later commit. At this point, it is still true I think 🤔

phip1611 replied:
You are right, I should rephrase the commit history in the early commits as we are now working towards a different goal!

phip1611 replied:
fixed!

amphi commented on vm-migration/src/progress.rs:
```diff
@@ -0,0 +1,424 @@
+}
+
+/// Carries information about the transmission of the VM's memory.
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    PartialOrd,
+    Ord,
+    PartialEq,
+    Eq,
+    Hash,
+    serde::Serialize,
+    serde::Deserialize,
+)]
```
Is there a reason to derive `Ord` or `Eq` when you never compare the complete struct? I'd expect that this is only necessary to compare this or to sort a vector of these structs.

phip1611 replied:
Personal preference to add them all to "basic" types.

amphi reviewed (commented):
Just a few nits, you have done great work here!

olivereanderson commented:
**Unimportant nit feel free to ignore :)** : I noticed that commit "9dd759c87a09e7113dc32ef75141620ca58a578d" includes "these information" in the commit message.  It should be "this information" as information is an uncountable noun and must thus always use the singular form.

Again not important and I don't really care as it is perfectly understandable.

olivereanderson commented on cloud-hypervisor/src/bin/ch-remote.rs:
```diff
@@ -524,8 +527,65 @@ fn rest_api_do_command(matches: &ArgMatches, socket: &mut UnixStream) -> ApiResu
+                    "migration progress should be there immediately when the migration was dispatched"
+                );
+
+                let progress = serde_json::from_slice::<MigrationProgress>(response.as_bytes())
+                    .map_err(|e| {
+                        error!("failed to parse response as MigrationProgress: {e}");
+                        Error::HttpApiClient(ApiClientError::ServerResponse(
+                            StatusCode::Ok,
+                            Some(response),
+                        ))
+                    })?;
+
```
Would it make sense to either log or `println!("{:?}", progress);` here, or will you already see everything that is relevant in the logs?

phip1611 replied:
good idea, it would make sense! Thanks

phip1611 replied:
Update: because of ` Some(response),` I will already see everything in the log

olivereanderson replied:
But only in the case of error right?

I was thinking to print/log the response regardless.

olivereanderson replied:
Or perhaps in the `Ongoing | Finished` match arms?

phip1611 replied:
But why? `ch-remote migration-progress` prints the entire JSON in the good case to the CLI

olivereanderson replied:
I see that this is more about ensuring that the call blocks until the migration is finished.

If anyone wants the migration statistics they can open a second `remote-ch` session that uses `migration-progress` 👍

olivereanderson reviewed (approved):
LGTM

Thanks for all the work you put into this important feature!

phip1611 committed:
ch-remote: wait for migration to finish by querying migration progress

We preserve the old behavior in ch-remote: SendMigration is blocking. A
new ´--dispatch` flag however ensures that one can just dispatch the
migration without waiting for it to finish (or fail).

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
ch-remote: add `migration-progress` command

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
vmm: migration: switch to non-blocking SendMigration call

Time has proven that the previous design was not optimal. Now, the
SendMigration call is not blocking for the duration of the migration.
Instead, it just triggers the migration. Using the new MigrationProgress
endpoint, management software can trigger the state of the migration and
also find information for failed migrations.

A new `keep_alive` parameter for SendMigration will keep the VMM alive
and usable after the migration to ensure management software can fetch
the final state. The management software is then supposed to send a
ShutdownVmm command.

With this, we are finally able to query the migration progress API
endpoint during an ongoing live migration.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
